runtime/src/main/kotlin/kotlin/reflect/AssociatedObjects.kt

```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+package kotlin.reflect
+
+import kotlin.native.internal.*
+
+@Retention(value = AnnotationRetention.BINARY)
+@Experimental(level = Experimental.Level.ERROR)
+annotation class ExperimentalAssociatedObjects
+
+@ExperimentalAssociatedObjects
+@Retention(AnnotationRetention.BINARY)
+@Target(AnnotationTarget.ANNOTATION_CLASS)
+annotation class AssociatedObjectKey
+
+@ExperimentalAssociatedObjects
+fun KClass<*>.getAssociatedObject(key: KClass<*>): Any? =
```

Comment:
That is somewhat un-Kotlin API (non typesafe & verbose). Public API should look something like

```
inline fun <reified T> KClass<*>.getAssociatedObject(): T
```

But even that way it does not look "native" with respect to the rest of reflection API (the rest of reflection consists mostly of vals). I wonder if @udalov can suggest something better.

Reply:
> `inline fun <reified T> KClass<*>.getAssociatedObject(): T`

Have you omitted `key` parameter intentionally?

Reply:
Writing `getAssiatedObject(Key::class)` looks less convenient to me than `getAssociatedObject<Key>()` (that's what refied type are parameters for)

Reply:
Ok. Then this function can't return `T`, because `T` is the annotation class then.
Maybe we should have `@AssociatedObject(KClass<*> key, KClass<*> value)` then instead.

Reply:
> I don't think that pursuing reified type parameter brings much value.

However it brings some of it, and it is more consistent with related reflection APIs.
I prefer using reified parameter here.

Reply:
Ok, we can start from the reified overload, and expose the non-reified one if there will be request for that.